src/passes/J2CLOpts.cpp

```diff
@@ -30,9 +30,56 @@ namespace wasm {
 
 namespace {
 
+using AssignmentCountMap = std::unordered_map<Name, Index>;
+using TrivialFunctionMap = std::unordered_map<Name, Expression*>;
+
 bool isOnceFunction(Function* f) { return f->name.hasSubstring("_<once>_"); }
 
-using AssignmentCountMap = std::unordered_map<Name, Index>;
+// Returns the function body if it is a trivial function, null otherwise.
+Expression* getTrivialFunctionBody(Module* m, Function* f) {
+  auto* body = f->body;
+  if (body->is<Return>()) {
+    body = body->dynCast<Return>()->value;
+  }
+  if (body == nullptr) {
+    // Empty body is trivial, use nop as the replacement for inlinling.
+    Builder builder(*m);
+    body = builder.makeNop();
+  }
```

Comment:
Lines 41-48 are covered by `remove-unused-brs` already, which is run elsewhere in this pass. Or does this need to happen at a higher frequency or something like that?

Reply:
This is also used to collect at the beginning where the cleanup has not been run  and I was not sure if we could rely that the ir was satisfying this invariant.

We could run the cleanup when collecting instead. Would that be your preference?

Reply:
Actually I still see `(return (global.get $$class@X))` after running the cleanup (`precompute`, `remove-unused-brs` and `vacuum`) 

Is that not expected?

Reply:
Is there anything else after the return? If there is then the return may be necessary. Otherwise it should definitely be removed, and that's worth looking into if you can point me to the testcase.

Reply:
Nope after running the three passes, the function ends up being just a `(return ...)`. And I can confirm that it is collected as a trivial function.  The only thing special is all those functions match the `noinline` pattern so when this pass ignores them they basically survive till the end and have a significant effect in codesize.

Reply:
Ok. Got to the bottom of this. It takes longer to make the function trivial because `vacuum`  is the pass that is converting something like

```
(block 
   (if .... )
   (nop)
   (global.get ...)
)
```

into 

```
 (return (global.get....)
```

And because we are running `remove-unused-brs`  before `vacuum`  it has no effects in these situations. Having `vacuum`  before `remove-unused-brs` makes the pass effective again.

Reply:
Uploaded a new patch that addresses this issue.

Reply:
That makes sense. There are some situations that require both passes in that order. The default pass pipeline runs them more than once for that reason. That should be very fast here, if we end up needing that.